README.md
# Homeyavr

Application to connect Homey to a Marantz AVR.
Version: 0.9.3

Currently working and supported in flows:

* power commands for AVR and main zone.
	- On
	- Off
* mute commands
	- On
	- Off
* source input selection
	- Select input source selections depending on the AVR type.
* volume
	 - Up
	 - Down
	 - Set (values between 0 - 80 )
* surround mode selection
	- Surround mode selections depending on the AVR type
* ECO depending on the AVR type
	- On
	- Off
	- Auto
* Triggers for:
	- power
	- mute
	- eco

* All selection strings and messages are using the "locales/<LANG>.json" files.
  I.e for new languages support only a new "locales/<LANG>.json" and app.json
  should be suffient.



App updates the internal status of the AVR constantly, even if the command is given
by a different application or by remote control, as long the AVR is transmitting
the commands.

Not working:
* Save settings.
  0.8.39 does not closed the settings window after save-settings and
  the changed values are not saved by Homey.
  Program will change the running parameters but after a restart the old values will be
  supplied by Homey and used.

To do:
* Use capabilities in the web interface as example light.
  Currently there are no capabilities defined (yet??) for the 'other' class.
  Changing the class to 'light' will give the on/off capability in the web interface.


Marantz AVR supported :
av8802, av8801, av7702, av7701, av7005,
sr7010, sr7009, sr7008, sr7007, sr7005,
sr6010, sr6009, sr6008, sr6007, sr6006, sr6005,
sr5010, sr5009, sr5008, sr5006, sr5005,
nr1606, nr1605, nr1604, nr1603, nr1602,
nr1505, nr1504


CHANGELOG.md

* V1.0.0
	- No changes: version used for first version to be uploaded into app-store.

* v0.9.4
	- Changes some print messages to debug messages in driver.js.

* V0.9.3
	- Add triggers from :
		- power
		- mute
		- eco

* v0.9.2
	- added jsonlint to Gruntfile
	- driver.js:
		- add event socket for communication between homey and avr controle part.
		- Changed device array
	- avr.js:
		- added a 'init' function.
		- added the event socket
		- define the AVR status requests functions as private.
	- conf json files:
		- use alphabetic order for inputsource and surround mode.
	- locales json files:
		- updated en.json and nl.json

* v0.9.1
	- Updated working.md file


* v0.9.0
	- Updated README.md
	- Updated / add pictures (svg & jpg)
	- driver.js:
		- settings function
	- conf json files: main_zone_power corrected

* v0.8.0
	- Re-write of early version.



src/test/avrsim.js
"use strict";

let net = require("net");

class AvrSim {

    constructor( sPort ) {

        this.port   = sPort;
        this.socket = null;

        this.powerstatus   = "PWSTANDBY";
        this.mzpowerstatus = "ZMOFF";
        this.mute          = "MUOFF";
        this.inputsel      = "SICD";
        this.volumestatus  = 45;
        this.surroundMode  = "MSAUTO";
        this.ecoMode       = "ECOON";
    }

    connect() {
        net.createServer ( (socket) => {
            this.socket = socket;

            socket
                .on( "listening", () => {
                    console.log(`AvrSim is listening on port ${this.port}.`);
                })

                .on( "data" , (data) => {
                    this.processData( data );
                })

                .on( "error" , (err) => {
                    console.log(`Error: ${err}.`);
                });
        }).listen( this.port );
    }

    close() {
        this.socket.destroy();
    }

    processData( data ) {
        let xData = String(data);

        xData = xData.replace("\r", "");

        let date = new Date();
        let dateStr = date.toISOString();
        console.log(`${dateStr} Received : '${xData}'.`);

        if ( xData.substr(0,2) === "PW") {
            if ( xData !== "PW?" ) {
                this.powerstatus = xData ;
            }
            console.log(`Returning : ${this.powerstatus}.`);
            this.socket.write( this.powerstatus);

        } else if ( xData.substr(0,2) === "ZM") {

            if ( xData !== "ZM?" ) {
                this.mzpowerstatus = xData ;
            }
            console.log(`Returning : ${this.mzpowerstatus}.`);
            this.socket.write( this.mzpowerstatus);

        } else if ( xData.substr(0,2) === "MU") {

            if ( xData !== "MU?" ) {
                this.mute = xData ;
            }
            console.log(`Returning : ${this.mute}.`);
            this.socket.write( this.mute);

        } else if ( xData.substr(0,2) === "SI") {

            if ( xData !== "SI?" ) {
                this.inputsel = xData ;
            }
            console.log(`Returning : ${this.inputsel}.`);
            this.socket.write( this.inputsel);

        } else if ( xData.substr(0,2) === "MV") {

            let re = /^MV(\d+)/i;

            if ( xData === "MVUP" ) {
                this.volumestatus++;

            } else if ( xData === "MVDOWN") {
                this.volumestatus--;

            } else {

                let Ar = xData.match(re);

                if ( Ar !== null ) {
                    this.volumestatus = Ar[1];
                    console.log(`Returning: MV${this.volumestatus}.`);
                }
            }
            console.log(`Returning : MV${this.volumestatus}.`);
            this.socket.write( "MV" + this.volumestatus );
        } else if ( xData.substr(0,2) === "MS") {

            if ( xData !== "MS?" ) {
                this.surroundMode = xData ;
            }
            console.log(`Returning : ${this.surroundMode}.`);
            this.socket.write( this.surroundMode);
        } else if ( xData.substr(0,2) === "EC") {

            if ( xData !== "ECO?" ) {
                this.ecoMode = xData ;
            }
            console.log(`Returning : ${this.ecoMode }.`);
            this.socket.write( this.ecoMode );
        }
    }
}

let avrSim = new AvrSim( 2222 );

avrSim.connect();


src/test/avrtest.js
/**
 * AVRTEST
 *      a basic node program to test the functionality of the 'avr.js' commands
 *      for a Marantz AVR remote commands over a telnet connetion.
 */

"use strict";

let IPADDRESS = "192.168.1.35";
let IPPORT    = 2222;

let Avr = require("./avr");

let mAvr = new Avr( IPPORT, IPADDRESS, "testavr", "SR5010");

console.log("IP address : ", mAvr.getHostname());
console.log("Port       : ", mAvr.getPort());
console.log("Type       : ", mAvr.getType());
console.log("Name       : ", mAvr.getName());

mAvr.setTest(); // enable testing (ignoring avr type restrictions).

/**
 * Check all power commands.
 *
 */
let checkPowerCommands = () => {

    setTimeout( () => {              mAvr.powerOn();             },   10);
    setTimeout( () => { console.log( mAvr.getPowerStatus());     }, 1000);
    setTimeout( () => { console.log( mAvr.getPowerOnOffState()); }, 1200);
    setTimeout( () => {              mAvr.powerOff();            }, 2000);
    setTimeout( () => { console.log( mAvr.getPowerStatus());     }, 2500);
    setTimeout( () => { console.log( mAvr.getPowerOnOffState()); }, 2700);
};

/**
 * Check all main zone power commands.
 *
 */
let checkMainZonePowerCommands = () => {

    setTimeout( () => {              mAvr.mainZonePowerOn();             },   10);
    setTimeout( () => { console.log( mAvr.getMainZonePowerStatus());     }, 1000);
    setTimeout( () => { console.log( mAvr.getMainZonePowerOnOffState()); }, 1200);
    setTimeout( () => {              mAvr.mainZonePowerOff();            }, 2000);
    setTimeout( () => { console.log( mAvr.getMainZonePowerStatus());     }, 2500);
    setTimeout( () => { console.log( mAvr.getMainZonePowerOnOffState()); }, 2700);
};

/**
 * Check all mute commands.
 *
 */
let checkMuteCommands = () => {

    setTimeout( () => {              mAvr.muteOn();             },   10);
    setTimeout( () => { console.log( mAvr.getMuteStatus());     },  500);
    setTimeout( () => { console.log( mAvr.getMuteOnOffState()); },  800);
    setTimeout( () => {              mAvr.muteOff();            }, 1000);
    setTimeout( () => { console.log( mAvr.getMuteStatus());     }, 1500);
    setTimeout( () => { console.log( mAvr.getMuteOnOffState()); }, 1800);
};

/**
 * Check all vlume commands.
 *
 */
let checkVolumeCommands = () => {

    setTimeout( () => {              mAvr.volumeUp();        },   10);
    setTimeout( () => { console.log( mAvr.getVolume());      },  500);
    setTimeout( () => {              mAvr.volumeDown();      },  800);
    setTimeout( () => { console.log( mAvr.getVolume());      }, 1300);
    setTimeout( () => {              mAvr.setVolume(33);     }, 1600);
    setTimeout( () => { console.log( mAvr.getVolume());      }, 2000);
};

/**
 * Check all input source commands.
 *
 */
let checkInputSourceCommands = () => {

    setTimeout( () => {              mAvr.selectInputSourcePhono();        },   10);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },  200);
    setTimeout( () => {              mAvr.selectInputSourceCd();           },  400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },  600);
    setTimeout( () => {              mAvr.selectInputSourceDvd();          },  800);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 1000);
    setTimeout( () => {              mAvr.selectInputSourceBluray();       }, 1200);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 1400);
    setTimeout( () => {              mAvr.selectInputSourceTv();           }, 1600);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 1800);
    setTimeout( () => {              mAvr.selectInputSourceSatCbl();       }, 2000);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 2200);
    setTimeout( () => {              mAvr.selectInputSourceSat();          }, 2400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 2600);
    setTimeout( () => {              mAvr.selectInputSourceMplay();        }, 2800);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 3000);
    setTimeout( () => {              mAvr.selectInputSourceVcr();          }, 3200);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 3400);
    setTimeout( () => {              mAvr.selectInputSourceGame();         }, 3600);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 3800);
    setTimeout( () => {              mAvr.selectInputSourceVaux();         }, 4000);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 4200);
    setTimeout( () => {              mAvr.selectInputSourceTuner();        }, 4400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 4800);
    setTimeout( () => {              mAvr.selectInputSourceSpotify();      }, 5000);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 5200);
    setTimeout( () => {              mAvr.selectInputSourceNapster();      }, 5400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 5600);
    setTimeout( () => {              mAvr.selectInputSourceFlickr();       }, 5800);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 6000);
    setTimeout( () => {              mAvr.selectInputSourceIradio();       }, 6200);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 6400);
    setTimeout( () => {              mAvr.selectInputSourceFavorites();    }, 6600);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 6800);
    setTimeout( () => {              mAvr.selectInputSourceAux1();         }, 7000);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 7200);
    setTimeout( () => {              mAvr.selectInputSourceAux2();         }, 7400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 7600);
    setTimeout( () => {              mAvr.selectInputSourceAux3();         }, 7800);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 8000);
    setTimeout( () => {              mAvr.selectInputSourceAux4();         }, 8200);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 8400);
    setTimeout( () => {              mAvr.selectInputSourceAux5();         }, 8600);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 8800);
    setTimeout( () => {              mAvr.selectInputSourceAux6();         }, 9000);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 9200);
    setTimeout( () => {              mAvr.selectInputSourceAux7();         }, 9400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            }, 9600);
    setTimeout( () => {              mAvr.selectInputSourceInetUsb();      }, 9800);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },10000);
    setTimeout( () => {              mAvr.selectInputSourceNet();          },10200);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },10400);
    setTimeout( () => {              mAvr.selectInputSourceBluetooth();    },10600);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },10800);
    setTimeout( () => {              mAvr.selectInputSourceMxport();       },11000);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },11200);
    setTimeout( () => {              mAvr.selectInputSourceUsbIpod();      },11400);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },11600);
    setTimeout( () => {              mAvr.sendInputSourceCommand("SICD");  },11800);
    setTimeout( () => { console.log( mAvr.getInputSelection());            },12000);
    setTimeout( () => { console.log( mAvr.getValidInputSelection());       },12200);
};

/**
 * Check all surround mode commands.
 *
 */
let checkSurroundCommands = () => {
    setTimeout( () => {              mAvr.setSurroundModeToMovies();         },    1);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                },  200);
    setTimeout( () => {              mAvr.setSurroundModeToMusic();          },  400);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                },  600);
    setTimeout( () => {              mAvr.setSurroundModeToGame();           },  800);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 1000);
    setTimeout( () => {              mAvr.setSurroundModeToDirect();         }, 1200);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 1400);
    setTimeout( () => {              mAvr.setSurroundModeToPureDirect();     }, 1600);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 1800);
    setTimeout( () => {              mAvr.setSurroundModeToStereo();         }, 2000);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 2200);
    setTimeout( () => {              mAvr.setSurroundModeToAuto();           }, 2400);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 2600);
    setTimeout( () => {              mAvr.setSurroundModeToNeural();         }, 2800);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 3000);
    setTimeout( () => {              mAvr.setSurroundModeToStandard();       }, 3200);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 3400);
    setTimeout( () => {              mAvr.setSurroundModeToDolby();          }, 3600);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 3800);
    setTimeout( () => {              mAvr.setSurroundModeToDts();            }, 4000);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 4200);
    setTimeout( () => {              mAvr.setSurroundModeToMultiChnStereo(); }, 4400);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 4600);
    setTimeout( () => {              mAvr.setSurroundModeToMatrix();         }, 4800);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 5000);
    setTimeout( () => {              mAvr.setSurroundModeToVirtual();        }, 5200);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 5400);
    setTimeout( () => {              mAvr.setSurroundModeToLeft();           }, 5600);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 5800);
    setTimeout( () => {              mAvr.setSurroundModeToRight();          }, 6000);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 6200);
    setTimeout( () => {              mAvr.sendSurroundCommand("MSAUTO");     }, 6400);
    setTimeout( () => { console.log( mAvr.getSurroundMode());                }, 6600);
    setTimeout( () => { console.log( mAvr.getValidSurround());               }, 6800);
};

/**
 * Check all eco commands.
 *
 */
let checkEcoCommands = () => {
    setTimeout( () => { console.log( mAvr.hasEco());              },                1);
    setTimeout( () => {              mAvr.ecoOn();                },              200);
    setTimeout( () => { console.log( mAvr.getEcoMode());          },              400);
    setTimeout( () => {              mAvr.ecoOff();               },              600);
    setTimeout( () => { console.log( mAvr.getEcoMode());          },              800);
    setTimeout( () => {              mAvr.ecoAuto();              },             1000);
    setTimeout( () => { console.log( mAvr.getEcoMode());          },             1200);
    setTimeout( () => {              mAvr.sendEcoCommand("ECO_UN_SUPPORTED"); }, 1400);
    setTimeout( () => { console.log( mAvr.getValidEcoCommands()); },             1600);
};

/**
 * Check all avr commands
 * Calls checks for power, mani zone power, mute, volume, inputsource, surround mode and eco.
 *
 */
let checkAll = () => {

    setTimeout( () => { checkPowerCommands();             },    10);
    setTimeout( () => { checkMainZonePowerCommands();     },  3000);
    setTimeout( () => { checkMuteCommands();              },  6000);
    setTimeout( () => { checkVolumeCommands();            },  8000);
    setTimeout( () => { checkInputSourceCommands();       }, 11000);
    setTimeout( () => { checkSurroundCommands();          }, 25000);
    setTimeout( () => { checkEcoCommands();               }, 35000);
    setTimeout( () => { process.exit(0);                  }, 40000);
};


checkAll();


avrsim/avrsim.js
"use strict";

let net = require("net");

class AvrSim {

    constructor( sPort ) {

        this.port   = sPort;
        this.socket = null;

        this.powerstatus   = "standby";
        this.mzpowerstatus = "off";
        this.mute          = "off";
        this.inputsel      = "CD";
        this.volumestatus  = 45;
    }

    connect() {
        net.createServer ( (socket) => {
            this.socket = socket;

            socket
                .on( "listening", () => {
                    console.log(`AvrSim is listening on port ${this.port}.`);
                })

                .on( "data" , (data) => {
                    this.processData( data );
                })

                .on( "error" , (err) => {
                    console.log(`Error: ${err}.`);
                });
        }).listen( this.port );
    }

    close() {
        this.socket.destroy();
    }

    processData( data ) {
        let xData = String(data);

        xData = xData.replace("\r", "");

        console.log(`Received : '${xData}'.`);

        if ( xData.substr(0,2) === "PW") {
            if ( xData === "PW?") {
                console.log(`Sending : ${this.powerstatus}.`);
                return this.powerstatus;
            } else {
                this.powerstatus = xData;
                console.log(`Returning: ${this.powerstatus}.`);
            }

        } else if ( xData.substr(0,2) === "ZW") {

            if ( xData === "ZW?") {
                console.log(`Sending : ${this.mzpowerstatus}.`);
                return this.mzpowerstatus;
            } else {
                this.mzpowerstatus = xData;
                console.log(`Returning: ${this.mzpowerstatus}.`);
            }

        } else if ( xData.substr(0,2) === "MU") {

            if ( xData === "MU?") {
                console.log(`Sending : ${this.mutestatus}.`);
                return this.mutestatus;
            } else {
                this.mutestatus = xData;
                console.log(`Returning: ${this.mutestatus}.`);
            }

        } else if ( xData.substr(0,2) === "SI") {

            if ( xData === "SI?") {
                console.log(`Sending : ${this.inputsel}.`);
                return this.inputsel;
            } else {
                this.inputsel = xData;
                console.log(`Returning: ${this.inputsel}.`);
            }

        } else if ( xData.substr(0,2) === "MV") {

            let re = /^MV(\d+)/i;

            if ( xData === "MV?") {
                console.log(`Sending : MV${this.volumestatus}.`);
                return this.volumestatus;
            } else if ( xData === "MVUP" ) {
                this.volumestatus++;
                console.log(`Returning: MV${this.volumestatus}.`);
            } else if ( xData === "MVDOWN") {
                this.volumestatus--;
                console.log(`Returning: MV${this.volumestatus}.`);
            } else {

                let Ar = xData.match(re);

                console.log( Ar );

                if ( Ar !== null ) {
                    this.volumestatus = Ar[1];
                    console.log(`Returning: MV${this.volumestatus}.`);
                }
            }
        }
    }
}

let avrSim = new AvrSim( 2222 );

avrSim.connect();


src/drivers/avr/lib/sr7007.js
"use strict";

let Avr = require("./avr");

let powerModes = [
    {
        type: "PWON",
        text: "power on"
    },
    {
        type: "PWSTANDBY",
        text: "power standby"
    }
];

let muteModes = [
    {
        type: "MUON",
        text: "Mute on"
    },
    {
        type: "MUOFF",
        text: "Mute off"
    }
];

let inputSources = [
    {
        type: "PHONO",
        text: "Phono"
    },
    {
        type: "CD",
        text: "CD player"
    },
    {
        type: "DVD",
        text: "DVD"
    },
    {
        type: "BD",
        text: "Bluray"
    },
    {
        type: "TV",
        text: "TV audio"
    },
    {
        type: "SAT/CBL",
        text: "Cable/Satelite"
    },
    {
        type: "SAT",       // Needs to be later than 'STA/CBL' for correct functioning.
        text: "Satelite"
    },
    {
        type: "MPLAY",
        text: "Media player"
    },
    {
        type: "GAME",
        text: "Game computer"
    },
    {
        type: "TUNER",
        text: "Tuner"
    },
    {
        type: "LASTFM",
        text: "LastFM"
    },
    {
        type: "FLICKR",
        text: "Flickr"
    },
    {
        type: "IRADIO",
        text: "Internet radio"
    },
    {
        type: "SERVER",
        text: "Server"
    },
    {
        type: "FAVORITES",
        text: "Favorites"
    },
    {
        type: "AUX1",
        text: "AUX1 port"
    },
    {
        type: "AUX2",
        text: "AUX2 port"
    },
    {
        type: "NET",
        text: "Network"
    },
    {
        type: "M-XPORT",
        text: "m-Xport"
    },
    {
        type: "USB/IPOD",
        text: "USB/IPOD port"
    }
];


class SR7007 extends Avr {

    /**
     * Constructor
     *
     * @method     constructor
     * @param      {number}  sPort  - The network port to be used.
     * @param      {string}  sHost  - The IP address of the AVR
     */
    constructor( sPort, sHost ) {
        super( sPort, sHost);

        this.powerstatus    = "unknown";
    }

    /*********************************************************************
     *  Power methods.
     *********************************************************************/

    /**
     * Power on a SR7007.
     *
     * @method     powerOn
     */
    powerOn() {

        let xThis = this ;

        this._writeCommandNoResponse("PWON");

        setTimeout( () => {
            xThis.getPowerStatusFromAvr() ;
        }, 2000);
    }

    /**
     * Power off / standby a SR7007
     *
     * @method     powerStandby
     */
    powerStandby() {

        let xThis = this ;

        this._writeCommandNoResponse("PWSTANDBY");

        setTimeout( () => {
            xThis.getPowerStatusFromAvr() ;
        }, 2000);
    }

    /**
     * Power off / standby a SR7007
     *
     * @method     powerStandby
     */
    powerOff() {

        let xThis = this ;

        this._writeCommandNoResponse("PWSTANDBY");

        setTimeout( () => {
            xThis.getPowerStatusFromAvr() ;
        }, 2000);
    }

    /**
     * Get the power status from the AVR.
     *
     * @method     getPowerStatusFromAvr
     */
    getPowerStatusFromAvr() {

        let xThis = this;

        this._writeCommandWithReponse("PW?", powerModes, xThis._cbPowerStatus.bind(xThis));
    }

    /**
     * Retunr the stored power status.
     *
     * @method     getPowerStatus
     * @return     {string}  - The stored power value.
     */
    getPowerStatus() {
        return this.powerstatus;
    }

    /*********************************************************************
     *  Mute methods.
     *********************************************************************/

    /**
     * Switch on mute
     *
     * @method     muteOn
     */
    muteOn() {

        let xThis = this ;

        this._writeCommandNoResponse("MUON");

        setTimeout( () => {
            xThis.getMuteStatusFromAvr() ;
        }, 1000);
    }

    /**
     * Switch off mute (unmute)
     *
     * @method     muteOff
     */
    muteOff() {

        let xThis = this ;

        this._writeCommandNoResponse("MUOFF");

        setTimeout( () => {
            xThis.getMuteStatusFromAvr() ;
        }, 1000);
    }

    /**
     * Get the mute status from the AVR.
     *
     * @method     getMuteStatusFromAvr
     */
    getMuteStatusFromAvr() {

        let xThis = this ;

        this._writeCommandWithReponse("MU?", muteModes,  xThis._cbMute.bind(xThis) );
    }

   /**
    * Return the stored mute status,
    *
    * @method     getMuteStatus
    * @return     {string}  - The stored mute status.
    */
    getMuteStatus() {
        return this.mute;
    }

    /*********************************************************************
     *  Inputsource methods.
     *********************************************************************/

   /**
    * Select PHONO as input source
    *
    * @method     selectInputSourcePhono
    */
    selectInputSourcePhono() {

        let xThis = this ;

        this._writeCommandNoResponse("SIPHONO");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select CD as input source
     *
     * @method     selectInputSourceCd
     */
    selectInputSourceCd() {

        let xThis = this ;

        this._writeCommandNoResponse("SICD");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

   /**
    * Select DVD as input source
    *
    * @method     selectInputSourceDvd
    */
    selectInputSourceDvd() {

        let xThis = this ;

        this._writeCommandNoResponse("SIDVD");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select Bluray as input source
     *
     * @method     selectInputSourceBluray
     */
    selectInputSourceBluray() {

        let xThis = this ;

        this._writeCommandNoResponse("SIBD");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select TV as input source
     *
     * @method     selectInputSourceTv
     */
    selectInputSourceTv() {

        let xThis = this ;

        this._writeCommandNoResponse("SITV");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

   /**
    * Select SAT/CBL as input source
    *
    * @method     selectInputSourceSatCbl
    */
    selectInputSourceSatCbl() {

        let xThis = this ;

        this._writeCommandNoResponse("SISAT/CBL");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select SAT as input source
     *
     * @method     selectInputSourceSat
     */
    selectInputSourceSat() {

        let xThis = this ;

        this._writeCommandNoResponse("SISAT");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

   /**
    * Select MPLAY as input source
    *
    * @method     selectInputSourceMplay
    */
    selectInputSourceMplay() {

        let xThis = this ;

        this._writeCommandNoResponse("SIMPLAY");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select GAME as input source
     *
     * @method     selectInputSourceGame
     */
    selectInputSourceGame() {

        let xThis = this ;

        this._writeCommandNoResponse("SIGAME");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select TUNER as input source
     *
     * @method     selectInputSourceTuner
     */
    selectInputSourceTuner() {

        let xThis = this ;

        this._writeCommandNoResponse("SITUNER");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select LASTFM as input source
     *
     * @method     selectInputSourceLastfm
     */
    selectInputSourceLastfm() {

        let xThis = this ;

        this._writeCommandNoResponse("SILASTFM");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select FLICKR as input source
     *
     * @method     selectInputSourceFlickr
     */
    selectInputSourceFlickr() {

        let xThis = this ;

        this._writeCommandNoResponse("SIFLICKR");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select Internet Radion as input source
     *
     * @method     selectInputSourceIRadio
     */
    selectInputSourceIRadio() {

        let xThis = this ;

        this._writeCommandNoResponse("SIIRADIO");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select SERVER as input source
     *
     * @method     selectInputSourceServer
     */
    selectInputSourceServer() {

        let xThis = this ;

        this._writeCommandNoResponse("SISERVER");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select Favorites as input source
     *
     * @method     selectInputSourceFavorites
     */
    selectInputSourceFavorites() {

        let xThis = this ;

        this._writeCommandNoResponse("SIFAVORITES");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

   /**
    * Select AUX1 as input source
    *
    * @method     selectInputSourceAux1
    */
    selectInputSourceAux1() {

        let xThis = this ;

        this._writeCommandNoResponse("SIAUX1");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select AUX2 as input source
     *
     * @method     selectInputSourceAux2
     */
    selectInputSourceAux2() {

        let xThis = this ;

        this._writeCommandNoResponse("SIAUX2");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

   /**
    * Select Net as input source
    *
    * @method     selectInputSourceNet
    */
    selectInputSourceNet() {

        let xThis = this ;

        this._writeCommandNoResponse("SINET");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select M-XPort as input source
     *
     * @method     selectInputSourceMXport
     */
    selectInputSourceMXport() {

        let xThis = this ;

        this._writeCommandNoResponse("SIM-XPORT");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Select USB/IPOD as input source
     *
     * @method     selectInputSourceUsbIpod
     */
    selectInputSourceUsbIpod() {

        let xThis = this ;

        this._writeCommandNoResponse("SIUSB/IPOD");

        setTimeout( () => {
            xThis.getInputSourceFromAvr() ;
        }, 1000);
    }

    /**
     * Get the input source selection from the AVR.
     *
     * @method     getInputSourceFromAvr
     */
    getInputSourceFromAvr() {
        let xThis = this ;
        this._writeCommandWithReponse("SI?", inputSources, xThis._cbInputSource.bind(xThis));
    }

   /**
    * Get the stored source input selection
    *
    * @method     getInputSource
    * @return     {string}  - The stored source input selection.
    */
    getInputSource() {
        return this.inputsource;
    }

    /*************************************************************************
     * Volume methodes
     *************************************************************************/

    /**
     * Increase the volume
     *
     * @method     increaseVolume
     */
    increaseVolume() {

        let xThis = this ;

        this._writeCommandNoResponse("MVUP");

        setTimeout( () => {
            xThis.getVolumeFromAvr() ;
        }, 1000);
    }

    /**
     * Descreas the volume
     *
     * @method     decreaseVolume
     */
    decreaseVolume() {

        let xThis = this ;

        this._writeCommandNoResponse("MVDOWN");

        setTimeout( () => {
            xThis.getVolumeFromAvr() ;
        }, 1000);
    }

    /**
     * Set the volime of the AVr (min:0, MAx: 80 )
     *
     * @method     setVolume
     * @param      {number}  level   The desired volume setting
     */
    setVolume( level ) {

        let xThis = this;

        if ( level >= 0 && level <= 80 ) {

            this._writeCommandNoResponse(`MV${level}`);

            setTimeout( () => {
                xThis.getVolumeFromAvr() ;
            }, 1000);
        }
    }

    /**
     * Get the volume setting from the AVR.
     *
     * @method     getVolumeFromAvr
     */
    getVolumeFromAvr() {

        let xThis = this ;

        this._writeCommandWithReponse("MV?", [] , xThis._cbVolume.bind(xThis));
    }

    /**
     * Get the store volume
     *
     * @method     getVolume
     * @return     {number}  - The stored last voluem settins.
     */
    getVolume() {
        return this.volume;
    }
}

module.exports = SR7007;
